### PR TITLE
feat(#204): added ctrl-j/k navigation hotkeys

### PR DESCRIPTION
**Pull request by AlexNabokikh**
## Summary
The purpose of this change is to assist the Vim community by enabling the use of ctrl-j/k for navigating through Ulauncher results. This combination is widely recognized as a navigation hotkey and is currently supported by most Linux, Mac, and Windows launchers.

## Local testing
The change has been tested successfully on Ubuntu 23.04.

### CODE REVIEW

**Comment by friday** [comment]
Thank you for this contribution. ctrl+j/k isn't really vim mode of course, but maybe it's close enough? Also see #909

**Comment by AlexNabokikh** [comment]
@friday I think It is close enough too. Lightweight Vim mode (ctrl-j/k) should suffice for the majority of users. To my knowledge, I haven't come across any launchers that support Vim's distinct modes (normal and insert) and functions. Instead, many popular launchers for Linux, Mac, Windows, and other applications simply incorporate ctrl-j/k/h/l.

**Review by friday — APPROVED** [review]

**Comment by troycurtisjr** [comment]
It is a pretty common workaround instead of implementing full modal
editing, however the trouble with this approach is that it conflicts with
the Ctrl-k keybinding when using the emacs keymap in Gnome. Ctrl-j is
technically a conflict as well, but seems highly improbable anyone would be
using it on a single line widget like with Ulauncher. Honestly even Ctrl-k
seems very uncommon for such a single-line mode, but it can still be useful
at times judging from personal use/experience.

Due to the known conflict, I'd highly appreciate this set of keymappings to
be optional...probably safely defaulted to on. Though I know that @friday
isn't super keen on more options 😀️

**Comment by AlexNabokikh** [comment]
@troycurtisjr Can you provide more details about the Emacs keymap in Gnome? Is it a non-standard mode? As a Gnome user, I have encountered several applications that use ctrl-k/j/h/l keys for navigation without any conflicts.

**Comment by friday** [comment]
I'm good with options that makes sense, such as keybindings. Can we make this simplified somehow so it's not a full key for key map settings page?

**Comment by AlexNabokikh** [comment]
@troycurtisjr Even after reading the article, I'm still confused about how to bump in keybinding overlap when the focused window is Ulauncher, but that may be just my lack of experience. 
Nevertheless, I believe that the Ulauncher community will benefit greatly by attracting Vim users. Many keyboard-centric users prefer launchers, and this is especially true for those who belong to the EMAC and VIM communities. I think, it would be fair to support such a large group such Vim users with this change.

**Comment by troycurtisjr** [comment]
Oh I agree and that is a major reason I put in the Ctrl-n/p key bindings 😁

In this case I'm just saying that we need a way to turn it off per user configuration because it is known to conflict with at least one power user oriented prevalent configuration, which as you point out is a demographic that launchers are popular with. 

Ulauncher is a gnome application, and as such unless explicitly configured otherwise by the application developer will automatically get all the text editing capabilities in the text input. When configured with the emacs key map, this means most of the editing shortcuts are also available. I'll note that this key map is also default in all major command line shells. If this is not made optional, then everyone with emacs key map will hit ctrl-k expecting to "kill line", but it will instead be an "up motion". Considering all the other line editing keys work (ctrl-b/a/f, alt-b, alt-backspace, etc) this singular conflict will greatly interrupt this category of power user flow.

This isn't just theoretical, it would immediately negatively impact my own usage. Hence my request for an option to disable this particular key binding.

**Comment by AlexNabokikh** [comment]
Ok. Any tips on how you want to make it optional? GUI settings checkmark, command line argument or any other options?

**Comment by AlexNabokikh** [comment]
@friday @troycurtisjr I added a check for Emacs keybindings, but I'm not sure if this helper function should go somewhere else.
Let me know if you have other ideas on how to do it.

**Comment by friday** [comment]
Looks like this solution could satisfy us all, assuming this works for Troy. Also I'm not sure ctrl+p/n supported is only needed for emacs mode @troycurtisjr? Should we disable that when we enable vim or is it good regardless?

Small nitpick. I would prefer something like this instead of duplicating the rest at the end:

```diff
+        up_aliases = ["p"]
+        down_aliases = ["n"]
+        if not self.is_emacs_bindings_active():
+            up_aliases = ["k", "p"]
+            down_aliases = ["j", "n"]

-            if keyname in ("Up", "ISO_Left_Tab") or (ctrl and keyname == "p"):
+            if keyname in ("Up", "ISO_Left_Tab") or (ctrl and keyname in up_aliases):
                self.results_nav.go_up()
                return True

-            if keyname in ("Down", "Tab") or (ctrl and keyname == "n"):
+            if keyname in ("Down", "Tab") or (ctrl and keyname in down_aliases):
                self.results_nav.go_down()
                return True
```

**Comment by friday** [comment]
Actually, when rethinking this I think the only good solution would be to add a dedicated preferences tab for keyboard shortcuts. The annoying thing about that is the current solution for setting the shortcut for opening Ulauncher uses an ugly convoluted solution to bring up a new GTK prompt (since the settings are otherwise a web app).

Might be time to abandon that solution, but pure html input fields doesn't let you set the modifier.

**Comment by troycurtisjr** [comment]
Yeah I agree that an explicit configuration is much better. I worry this is trying to be too clever and makes too many assumptions about what the user wants. I think we can just include an option in the file to disable these shortcuts. Just a pin of a discussion or similar on updating the setting is likely enough, as I do think it will be a pretty small minority case (even if it does include me). If we notice it being used more, perhaps for some other reason than the Gnome emacs keymap scenario, then we can add it to the UI, or we can add it as part of the more comprehensive keymap configuration @friday mentioned.

**Comment by AlexNabokikh** [comment]
@troycurtisjrI modified the code to only verify the specific parameter in settings.json. I'm unsure of how to incorporate it into the GUI menu or if it's necessary. Perhaps it can be noted in the documentation instead, just like you mentioned.

**Comment by friday** [comment]
That doesn't fully solve the problem imo. The problem is that all of these vim, emacs etc allow customization.

How about changing the setting name to `arrow_key_aliases` (string) and expect it to hold 4 characters (like `"hjkl"`). Then in the code, validate that the string is exactly 4 characters long, and if so, grab the aliases:

```py
arrow_key_aliases_valid = len(self.settings.arrow_key_aliases) == 4
arrow_key_aliases = [*self.settings.arrow_key_aliases] if arrow_key_aliases_valid else [None] * 4
left_alias, down_alias, up_alias, right_alias = arrow_key_aliases
```

Then you can compare directly with those (change `in up_aliases` to `= up_alias` and same for down)

And also add `and keyname` to the condition `if self.results_nav:` to be sure it's not `None` (I'm not sure if it can be None if you just press a modifier or the input is invalid, non-standard keyboard layout or something).

We don't actually use the left/right aliases for anything, but we really should. The Gtk (text) Entry takes care of left/right for us, but it's definitely possible to add aliases for it on top of those. If you think this is something you want to do, feel free, otherwise just leave that part (change the unused variable names to _left_alias, and _right_alias).

Basically this would be just like the like "jump keys" setting, and similar to them you can't set the modifier, just the key. Long term I think we want more advanced settings, but for now I think this is as flexible as we can possibly go.

I already had you revise this solution a lot, so I would recommend to wait for @troycurtisjr to have a say on this too in case he disagrees.

**Comment by friday** [comment]
I don't know what the most logical order would be for the sting, so I just went with the vim "hjkl" order. If gamers want to use "wasd" they would have to put it in like "aswd" for example. In the preferences web view you could duplicate the jump_keys section and change the key and description, but again that's optional (easy for me to add afterwards).

**Comment by AlexNabokikh** [comment]
@troycurtisjr Can you please provide your opinion on that?

**Comment by troycurtisjr** [comment]
Sorry it took so long to get back around to this. 

I think this PR makes a good incremental positive progress. The more complete configuration which @friday described is certainly more capable, but it doesn't seem strictly necessary. If it is easy enough to add in, then it does give the maximum flexibility. Since the default should likely be "hjkl" that seems like a reasonable order to go with. We can always make the GUI more explicit with broken out fields so there is no confusion around the ordering. I also agree the GUI part is not required at this point.

So basically I think either way works, the PR as is, or the more complete option laid out by @friday .

**Comment by friday** [comment]
@AlexNabokikh I pushed some changes to speed it up a little. Does this work for you?

**Comment by AlexNabokikh** [comment]
@friday totally! Looks great!

**Comment by friday** [comment]
I just merged #1256, so if you upgrade now the hotkey support will be using a completely different solution, with different support. If you are using Gnome or if you already set it up with your DE before (as per the recommendations to support Wayland) you should see no change. If you are on KDE or XFCE I added those too but haven't tested myself yet.

If not, then you will have to set it up with your DE settings. You should get a message prompt to help.

**Comment by friday** [comment]
And @troycurtisjr you will have to change/add your settings.json arrow_key_aliases now to some four letter combination like "xnpz"
